### PR TITLE
feat: add cleavejs formatting

### PR DESCRIPTION
**Pull request by jason-capsule42**
# Alaska Airlines Pull Request

**Fixes:** #8

## Summary:

This change adds CleaveJS support for input formatting. All formatting options supported in CleaveJS are supported here. This change also includes credit card icons for inputs formatted for CC.

## Type of change:

Please delete options that are not relevant.

- [x] New capability
- [ ] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

File: src/auro-input.js
Comment (gusnaughton) [inline src/auro-input.js:158]: For the future, it would be really handy to know where these magic numbers come from

File: demo/index.html
Comment (blackfalcon) [inline demo/index.html:106]: This is an interesting bonus, but for the final demo we will not promote support to usage at this time until we have completed other work that is in design.

File: demo/index.html
Comment (blackfalcon) [inline demo/index.html:112]: Same...

File: demo/index.html
Comment (blackfalcon) [inline demo/index.html:118]: Same...

File: demo/index.html
Comment (blackfalcon) [inline demo/index.html:124]: Same...

File: demo/index.html
Comment (blackfalcon) [inline demo/index.html:130]: Same...

File: demo/index.html
Comment (blackfalcon) [inline demo/index.html:136]: Same...

File: src/auro-input.js
Comment (blackfalcon) [inline src/auro-input.js:125]: This is a private method. Please add the `@private` JSDoc comments.

File: src/auro-input.js
Comment (blackfalcon) [inline src/auro-input.js:135]: This is a private method. Please add the `@private` JSDoc comments.

File: src/auro-input.js
Comment (blackfalcon) [inline src/auro-input.js:221]: This is a private method. Please add the `@private` JSDoc comments.

File: src/base-input.js
Comment (blackfalcon) [inline src/base-input.js:88]: These properties were grouped by Sting then Boolean. Wondering why these were places here?

File: demo/index.html
Comment (blackfalcon) [inline demo/index.html:100]: This demo confuses me? For the format, `icon` is `true`, but this is the no-icon version?

File: src/auro-input.js
Comment (blackfalcon) [inline src/auro-input.js:208]: This could use comments or possibly be higher up in the code. It took me a long time to find this as the hook for showing the icon or not. There is a lot of code for this feature so some comments on how this all ties together would be very helpful.

File: src/auro-input.js
Comment (blackfalcon) [inline src/auro-input.js:136]: Is there to be a comment here? `//`

**Review by blackfalcon — CHANGES_REQUESTED** [review]

File: demo/index.html
Comment (jason-capsule42) [reply demo/index.html:100]: The key should have been removed when I switched course to use the property but was overlooked. Updated the PR.

File: src/base-input.js
Comment (jason-capsule42) [reply src/base-input.js:88]: Moved to after list of strings

File: src/auro-input.js
Comment (jason-capsule42) [reply src/auro-input.js:221]: Added private comment

File: src/auro-input.js
Comment (jason-capsule42) [reply src/auro-input.js:208]: Added private comment

File: src/auro-input.js
Comment (jason-capsule42) [reply src/auro-input.js:136]: Nope, cleanup artifact. Removed.

File: src/auro-input.js
Comment (jason-capsule42) [reply src/auro-input.js:158]: Good call out @gusnaughton - adding a link in a comment at the top of the method to https://www.creditcardinsider.com/learn/anatomy-of-a-credit-card/#credit-card-number

File: src/auro-input.js
Comment (gusnaughton) [inline src/auro-input.js:129]: Would it be possible to also support the Alaska card logo for Alaska issued Visa Signature cards? Additionally, in the future, commercial accounts and e-certificate codes?

Comment (gusnaughton) [reply src/auro-input.js:129]: The payment methods available on the Booking Payment page would be a great example of this

Comment (blackfalcon) [reply src/auro-input.js:129]: Does it appear from the code that we are just looking for signature prefix numbers? I would assume that the Alaska card will have a signature set of numbers we can filter on? Cause yeah, we NEED to show the Alaska card icon. 

Good catch @gusnaughton

Comment (jason-capsule42) [reply src/auro-input.js:137]: Cards starting in "1" or "2" signifies a card associated with a "Major Industry Identifier (MII)" for Airlines. Digits 2-6 identify a particular institution.

From this I would gather it's likely that the AlaskaAir credit would have its on 2-6 digit identifier at the start of every card number.

https://www.creditcardinsider.com/learn/anatomy-of-a-credit-card/#credit-card-number

Adding another identifier to show a custom card for any identifier is trivial - given that unique identifier. @gusnaughton @blackfalcon either of you able to point me to where I can get that identifier value?

Comment (jason-capsule42) [reply src/auro-input.js:138]: Added the matched values for Alaska Air credit cards.

File: docs/api.md
Comment (blackfalcon) [inline docs/api.md:5]: Be sure to commit updated API doc. This is not done as part of the release process.

File: src/auro-input.js
Comment (blackfalcon) [inline src/auro-input.js:228]: There is no test for this scenario.

**Review by blackfalcon — CHANGES_REQUESTED** [review]

File: src/auro-input.js
Comment (geoffrich) [inline src/auro-input.js:156]: Can we extract the icon determination into a separate utility function? It's quite long.

File: test/auro-input.test.js
Comment (geoffrich) [inline test/auro-input.test.js:221]: I wouldn't check to.be.accessible on every test. These tests should focus on setting the right card icon.

File: test/auro-input.test.js
Comment (geoffrich) [inline test/auro-input.test.js:223]: Can we move this query inside the setInputValue helper? That way we don't have to do it in every test.

**Review by geoffrich — CHANGES_REQUESTED** [review]

File: test/auro-input.test.js
Comment (jason-capsule42) [reply test/auro-input.test.js:221]: Good point - removing this.

File: test/auro-input.test.js
Comment (jason-capsule42) [reply test/auro-input.test.js:223]: This is good thinking. Also going to simplify the code for cc card icon validation into a function rather than repeat it in each test.

File: src/auro-input.js
Comment (jason-capsule42) [reply src/auro-input.js]: This got me thinking and I decided to change this away from the overly complicated switch statement and move it to a simple regex checker.

Much cleaner! Thanks for instigating @geoffrich.

File: docs/api.md
Comment (jason-capsule42) [reply docs/api.md:5]: Thanks, updated.

File: src/auro-input.js
Comment (blackfalcon) [inline src/auro-input.js:30]: Please add `@private` to this as well. This does not need to be in the public API.

File: src/base-input.js
Comment (blackfalcon) [inline src/base-input.js:88]: I'd argue that this is not to be in the public API? This object is only for internal use, right? Is there a use case when someone would pass that in? 

Do we need to set this property?

**Review by blackfalcon — CHANGES_REQUESTED** [review]
This is looking really good! Just a couple of final questions on API and we are golden.

File: src/auro-input.js
Comment (jason-capsule42) [reply src/auro-input.js]: Test for AlaskaAir card is now added

File: src/auro-input.js
Comment (jason-capsule42) [reply src/auro-input.js:33]: Done

File: src/base-input.js
Comment (jason-capsule42) [reply src/base-input.js:88]: This works only because it's passed in. Every example in the doc passes in a format JSON blob. In some cases it's as simple as `"{"creditCard": true}"` but it could be far more complex if the implementer wants. They can pass in any JSON object supported by CleaveJS to format for dates, phone numbers or anything else they may want.

Comment (blackfalcon) [reply src/base-input.js:88]: Sorry, I should have been more clear. Since the only format we are releasing is the credit card, I am wondering if this can be an internal setting. All the other features are TBD as to when they will be released.  

When we get to things like `format='{"date": true, "delimiter": "-", "datePattern": ["Y", "m", "d"]}` I would want to be more controlled in how this API works as there are specific formats we are to support per design and UX writing rules. https://auro.alaskaair.com/components/auro/datetime

I see the value in the open-ended API for cleave, but for our purposes, we will want this to be more specific.

Comment (jason-capsule42) [reply src/base-input.js:88]: I figured out a pretty simple way to restrict the format to only apply if it's set to be a credit card. Essentially, just calling cleaveJS if the formatJSON has the `creditCard` key in it. Doing it this way means enabling other formats in the future would preserve a consistent approach to defining the format rather than credit cards having a custom wrapper layer. 

I removed the other format examples from the demo page as well.

**Review by geoffrich — COMMENTED** [review]
Apologies for not bringing this up in my first review -- I hadn't fully processed the implications of passing a Cleave format object in the API.

I'm not a fan of exposing the entire Cleave object to the API. I would rather abstract it away behind something like format="creditcard" and have the component pass the relevant config. This is more restrictive, yes, but intentionally so.

- the current implementation allows the consumer to pass anything. Yes, we're checking that credit card is in the object, but they can still pass any of the options in the [general config](https://github.com/nosir/cleave.js/blob/master/doc/options.md#general-config). This increases the API surface we have to maintain. Validating this object is going to be difficult as we allow other options. I would prefer mapping the format property to a Cleave config in the component implementation, e.g. 
```js
if (format === 'creditcard') {
	config = { creditCard: true };
} // when we add more options
else if (format === 'date') {
	config = { date: true, datePattern: ['Y', 'm', 'd'], delimeter = '-' };
} // etc
```
- by making this customizable, there's potential for design inconsistency, especially as we add other formats. For instance, I don't think the consumer should be able to customize the [delimeter](https://github.com/nosir/cleave.js/blob/master/doc/options.md#delimiter) used -- this should be consistent across all uses of the component. In its current state, we either have to allow everything or try to validate certain options, which will get messy. It's much simpler if we have full control over the objects passed to Cleave.
- using an object here makes the API more complex. As a consumer, I would prefer a simple option like format="creditcard" as opposed to remembering the format of the Cleave object. The consumer of this component should not care that we're using Cleave.
- we're also taking a hard dependency on Cleave, or at least its config object. If we ever replace the dependency, we would either need to replicate the Cleave API (since the consumer could be passing anything) or force a breaking change. By abstracting this behind a simpler format option, we can swap out the dependency any time we like as long as the formats we explicitly support still work.

**Review by blackfalcon — CHANGES_REQUESTED** [review]
We have a significant miss. When the input for CC appears on a mobile device, the standard text entry keyboard appears. Couple that with cleave not allowing alpha characters, the keyboard appears broken. 

This feels like a good option to correct;
https://css-tricks.com/finger-friendly-numerical-inputs-with-inputmode/#a-better-solution-inputmode

Adding the `inputmode` solves this issue, but we can't have that there all the time. So this attribute needs to be bound to something else. I have concerns about binding this additional attribute to an open object entry being used to use the CC feature. 

From a user's perspective, it needs to be declarative that this will be the CC version of the input.

**Comment by blackfalcon** [comment]
> I'm not a fan of exposing the entire Cleave object to the API. I would rather abstract it away behind something like format="creditcard" and have the component pass the relevant config. This is more restrictive, yes, but intentionally so.

```js
if (format === 'creditcard') {
	config = { creditCard: true };
} // when we add more options
else if (format === 'date') {
	config = { date: true, datePattern: ['Y', 'm', 'd'], delimeter = '-' };
} // etc 
```

For additional context, another team is picking up work regarding the use of a date-entry input field. The coming consensus is that the date **MUST** adhere to the standard that the [auro-datetime](https://auro.alaskaair.com/components/auro/datetime) element is using for a numeric date. 

Leaving this open-ended with direct access to the cleave API will allow users to use unsupported formats.

**Comment by jason-capsule42** [comment]
@geoffrich @blackfalcon I've made two notable updates to the PR.

1. The config for formatting is changed to pass in only a string to designate which format to use. Based on that string value, and only if the string value matches an expected option, the correct JSON config is defined and passed into CleaveJS. Currently only `credit-card` is supported. All other strings passed in will be ignored and CleaveJS will not instantiate.

2. If the format is `credit-card` the attributes to ensure mobile devices will use a numeric keyboard are added to the input element.

File: src/auro-input.js
Comment (blackfalcon) [inline src/auro-input.js:151]:
```suggestion
    return this.icon && this.format === 'credit-card'
```

I am pretty sure that we don't need the additional validation of `this.format`.  This would be true for any of these in this doc.

File: src/base-input.js
Comment (blackfalcon) [inline src/base-input.js:89]: Had a side chat with @geoffrich regarding the forming [API standard](https://github.com/AlaskaAirlines/WC-Generator/issues/105) and we can't come up with a good reason to NOT extend the native HTML spec. 

Instead of creating a new attribute for this, why not extend the `type` attribute? Instead of describing a new attribute in the API, we simply add another option to the already existing `text`, `password`, etc...  options.

To that end, I think we have the opportunity to redefine the UI for `type="date"` when we get the green light on the UI for this. Given the designs we are seeing, we will never be using the native `type="date"`.

File: demo/index.html
Comment (blackfalcon) [inline demo/index.html:94]:
```suggestion
          <auro-input id="format-ccWithIcon" format="credit-card" icon label="Credit Card Number with Icon" required></auro-input>
```

HTML boolean attributes do not have a `true` or `false` value. 

`icon=""` === true
`icon="icon"` === true
`icon="false" === true

We can't assume that `icon="true"` is `true` because `icon="false"` will return true. Only the absence of the attribute is false.

https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes

This is not to be confused with Aria attributes that are not boolean, but support a [tristate](https://www.w3.org/TR/wai-aria-1.1/#valuetype_tristate) value.

**Review by blackfalcon — CHANGES_REQUESTED** [review]

**Comment by blackfalcon** [comment]
> @geoffrich @blackfalcon I've made two notable updates to the PR.
> 
> 1. The config for formatting is changed to pass in only a string to designate which format to use. Based on that string value, and only if the string value matches an expected option, the correct JSON config is defined and passed into CleaveJS. Currently only `credit-card` is supported. All other strings passed in will be ignored and CleaveJS will not instantiate.
> 2. If the format is `credit-card` the attributes to ensure mobile devices will use a numeric keyboard are added to the input element.

These updates look great. Made some suggestions about binding this API to the native `type` API. 

Also, not seeing any updates regarding the issue with the input and the keyboard on a mobile device?

File: src/auro-input.js
Comment (jason-capsule42) [reply src/auro-input.js:151]: In VueJS I've always had to be explicit in this way so it became my habit to always do the pre-check for the attribute before the comparison. Looks like litElement doesn't complain in the same way so I'll remove the instances of this.

File: src/base-input.js
Comment (jason-capsule42) [reply src/base-input.js]: I like that idea. I patched in a change - check it out and let me know if that matches up to what you are thinking.

**Comment by jason-capsule42** [comment]
> > @geoffrich @blackfalcon I've made two notable updates to the PR.
> > 
> > 1. The config for formatting is changed to pass in only a string to designate which format to use. Based on that string value, and only if the string value matches an expected option, the correct JSON config is defined and passed into CleaveJS. Currently only `credit-card` is supported. All other strings passed in will be ignored and CleaveJS will not instantiate.
> > 2. If the format is `credit-card` the attributes to ensure mobile devices will use a numeric keyboard are added to the input element.
> 
> These updates look great. Made some suggestions about binding this API to the native `type` API.
> 
> Also, not seeing any updates regarding the issue with the input and the keyboard on a mobile device?

I made an update to use type as suggested. Great idea. I like it.

re: keyboard input on mobile - double check, you must have just overlooked it.

```
        inputmode="${ifDefined(this.numericKeyboard ? `numeric` : undefined)}"
        pattern="${ifDefined(this.numericKeyboard ? `[0-9]` : undefined)}"
```

File: src/auro-input.js
Comment (geoffrich) [inline src/auro-input.js:232]: Do we still need this eslint disable?

File: src/auro-input.js
Comment (geoffrich) [inline src/auro-input.js:241]: Do we still need this eslint disable?

File: src/base-input.js
Comment (geoffrich) [inline src/base-input.js:26]:
```suggestion
 * @attr {Boolean} icon - If set, will render an icon inside the input to the left of the value. Support is limited to auro-input instances with type credit-card.
```

File: src/auro-input.js
Comment (geoffrich) [inline src/auro-input.js:255]: I don't think we need ifDefined here, since undefined is already the else case in the ternary

File: test/auro-input.test.js
Comment (geoffrich) [inline test/auro-input.test.js:291]: We don't need to run this on the icon -- I don't think axe will ever flag an a11y violation here.

File: test/auro-input.test.js
Comment (geoffrich) [inline test/auro-input.test.js:292]: Does this expect need to be awaited? If not, we can remove it and this whole function can be synchronous.

**Review by geoffrich — CHANGES_REQUESTED** [review]
See my comments for a few nitpicks.

More importantly -- this PR appears to have broken email validation. Steps to repro:
1. Load the demo page locally
2. Navigate to the email input (/demo/#emailTest01)
3. Type a single character into the field
4. Blur the field. No error message shows. The email input on the [doc site](https://auro.alaskaair.com/components/auro/input) does show an error.

Also, my cursor gets moved when typing into the credit card field. See gif below
ccinput

These two issues need to be resolved before merging.

File: docs/api.md
Comment (blackfalcon) [reply docs/api.md:11]: FYI... I fixed this in the `next` branch for the WC-Generator. It's more automated.

File: test/auro-input.test.js
Comment (jason-capsule42) [reply test/auro-input.test.js:292]: This does need to be awaited. The tests fail without it when entering values that cause the card icon to update.

File: src/auro-input.js
Comment (jason-capsule42) [reply src/auro-input.js:253]: When ifDefined is removed than all inputs that don't use the numeric keyboard have `inputmode="undefined" pattern="undefined"`. When using ifDefined the attributes are not added at all.

**Comment by geoffrich** [comment]
2 new issues I found (sorry Jason):

- When I enter a credit card number the validation message "Please match the requested format" appears. I assume this is because we set `pattern="[0-9]"` and cleave is inserting spaces. Maybe the pattern regex needs to include a space? Found in Chrome and Firefox
- There is no limit on the credit card input length in Firefox (see below screenshot)

image

**Review by blackfalcon — CHANGES_REQUESTED** [review]
This is all looking really good! I updated the demo page to really go over all the options of this element. There are a lot and it's confusing. 

Any issue with the type="email" appears to be addressed. The feature works consistently here as it does on the doc site. 

There is one last issue I found. The error state for credit cards exposed an issue that the credit-card icon placement moves with an error or if there is placed help text.

Screen Shot 2021-05-01 at 3 08 41 PM

Is there anything in the design about the disabled state? It's working, but the icon is solid black versus the disabled gray. I am not sure what the intention is here. 

Screen Shot 2021-05-01 at 3 14 26 PM

File: src/base-input.js
Comment (blackfalcon) [inline src/base-input.js:24]:
```suggestion
 * @attr {String} type - Populates the `type` attribute on the input. Allowed values are `password`, `email`, `credit-card`  or `text`. If given value is not allowed or set, defaults to `text`.
```

File: src/style.scss
Comment (blackfalcon) [inline src/style.scss:201]: Positioning needs to be adjusted to support when helptext is added or there is an error state.

File: src/auro-input.js
Comment (blackfalcon) [inline src/auro-input.js:185]: Does Cleave have a way to validate if all the numbers have been entered? Just noticed that there is no validation check being done on blur from the input. 

If Cleave can support this, is there a message that can be placed in the `error` attribute?

**Review by blackfalcon — CHANGES_REQUESTED** [review]

File: src/auro-input.js
Comment (jason-capsule42) [reply src/auro-input.js:185]: CleaveJS does not handle any form of validation - purely formatting. We would need to extend our own validation rules here.

This is good thinking, @blackfalcon. I'll get a check for that baked in.

**Comment by jason-capsule42** [comment]
Latest update includes custom validation message when not entering the required number of digits for a credit card and improves positioning of icon when an error message is displayed.

**Review by blackfalcon — APPROVED** [review]

**Pull request merged**
